### ISSUE TITLE
Need to improve print_control_identifiers() performance with depth param

### ISSUE THREAD

**Issue by vasily-v-ryabov** [pywinauto/pywinauto#359]
I've observed very slow behavior for iTunes (tab "Store").
```python
>>> len(app.iTunes.descendants()) # takes ~2.5-3 sec.
3024
>>> len(app.iTunes.children()) # takes less than 0.1 sec.
5
```

`app.iTunes.print_control_identifiers()` takes more than a minute! Couldn't wait it for a reasonable time. Even if I set argument `depth=1`.

I think we need to improve this at least for some constrained depth value.

**Comment by cetygamer** [comment]
The cause is `UIAWrapper.__hash__` method.
https://github.com/pywinauto/pywinauto/blob/master/pywinauto/controls/uiawrapper.py#L198
A lot of controls don't have RuntimeId field, so mapping dictionary out of controls in order to build control tree takes a very long time.
Typical item in iTunes app looks like this:
![image](https://cloud.githubusercontent.com/assets/14858753/26284714/76998ad0-3e4a-11e7-8dc0-dbc91d305881.png)

**Comment by airelil** [comment]
Should we consider to generated our own unique 'hash' value in this case ?

**Comment by nimuston** [comment]
Hi,

Could this be also solution for #256 issue? i'm still struggling with this slowness problem

**Comment by airelil** [comment]
@nimuston, you can easily check it with a small trick.  See an example below:
```python
import pywinauto
import pywinauto.controls.uiawrapper

 def hash_stub(self):
     return 21   # return any number, we don't care for now

pywinauto.controls.uiawrapper.UIAWrapper.__hash__ = hash_stub

app = pywinauto.application(backend='uia')
app.connect(path="itunes.exe')   # check for the exact name here...
w = app.window(title_re=".*itunes.*")
w.print_control_identifiers()

```

**Comment by nimuston** [comment]
with that example it takes very long time to print identifiers. Longer than usual.

**Comment by vasily-v-ryabov** [comment]
Still prefer to keep this issue open. Need further efforts to optimize the algorithm.

**Comment by nimuston** [comment]
i also tested how the new fix works in my case. Printing identifiers is almost 1 min faster, so great improvement. Is it now using the same logic also to find actual buttons or is this fix only for printing identifiers?

**Comment by vasily-v-ryabov** [comment]
@nimuston it may affect on search engine as well because it uses updated operator `__eq__` to compare elements (the rest of fixes should impact on printing ids only). Not sure the impact on search engine is big enough. Had no chance to test it deeper yet.

**Comment by nimuston** [comment]
@vasily-v-ryabov what do you think, is it possible to make search engine work faster?

**Comment by vasily-v-ryabov** [comment]
Hope it's possible. Algorythm complexity looks non-linear on number of controls (maybe quadratic or 3-polynomial). We're planning to continue investigation as I said above.

**Comment by vasily-v-ryabov** [comment]
This issue as it was submitted initially is fixed in `atspi` branch by PR #1049. Let's continue other optimization topics in #256.
